lib/vector-core/src/fanout.rs

```diff
@@ -42,13 +37,81 @@ impl Fanout {
 
         let fanout = Self {
             sinks: vec![],
-            i: 0,
             control_channel: control_rx,
         };
 
         (fanout, control_tx)
     }
 
+    /// Send a full vec of Events into the Fanout
+    ///
+    /// This function accepts a `Vec<Event>` and duplicates it down into the
+    /// Fanout's sinks. We take care to avoid the slow receiver problem as much
+    /// as possible but any misbehaving downstream WILL cause this function to
+    /// suffer high latencies. The memory use of this function is
+    /// `N*mem::size_of(events)`, where N is the number of downstream targets.
+    ///
+    /// # Errors
+    ///
+    /// Function will error with `()` in case downstream Sink operations
+    /// fail. This is, we recognize, not very helpful.
+    ///
+    /// # Panics
+    ///
+    /// This function will only panic if there are internal logic errors,
+    /// specifically around the size of the buffer created to house clones of
+    /// Vec<Event>.
+    pub async fn send_all(&mut self, events: Vec<Event>) -> Result<(), ()> {
+        self.process_control_messages().await;
+
+        if self.sinks.is_empty() {
+            // nothing, intentionally
+        } else if self.sinks.len() == 1 {
+            if let Some(sink) = &mut self.sinks[0].1 {
+                for event in events {
+                    sink.feed(event).await?;
+                }
+                sink.flush().await?;
+            }
+        } else {
+            let mut faulty_sinks = Vec::new();
+
+            {
+                let mut jobs = FuturesUnordered::new();
+                let mut clone_army: Vec<Vec<Event>> =
+                    Vec::with_capacity(self.sinks.iter().filter(|x| x.1.is_some()).count());
+                for _ in 0..(self.sinks.len() - 1) {
+                    clone_army.push(events.clone());
+                }
```

Comment:
This is still wrong. The problem is in the push, where you only need to push if the sink is not `None`. This just reduces the capacity of the initial `Vec` and will then require a reallocation. Maybe this?
```suggestion
                let count = self.sinks.iter().filter(|x| x.1.is_some()).count();
                let mut clone_army: Vec<Vec<Event>> = Vec::with_capacity(count);
                for _ in 0..(count - 1) {
                    clone_army.push(events.clone());
                }
```